api/agent/pure_runner.go

```diff
@@ -836,87 +831,78 @@ func (pr *pureRunner) runStatusCall(ctx context.Context) *runner.RunnerStatus {
-		isWaiter = true
-		myChan = pr.status.wait
-	} else {
-		// Wait channel is not present, we install a new one.
-		myChan = make(chan struct{})
-		pr.status.wait = myChan
+	isCached = now.Before(pr.status.expiry)
+	if !isCached {
+		if pr.status.wait == nil {
+			pr.status.wait = make(chan struct{})
+			isSpawner = true
+		}
+		waitChan = pr.status.wait
 	}
 
 	pr.status.lock.Unlock()
 
-	// We either need to wait and/or serve the request from cache
-	if isWaiter || isCached {
-		if isWaiter {
-			select {
-			case <-ctx.Done():
-				return nil, ctx.Err()
-			case <-myChan:
-			}
-		}
+	if isSpawner {
+		go func() {
+			// IMPORTANT: We have to strip client timeouts to make sure this completes
+			// in the background even if client cancels/times out.
+			cachePtr := pr.runStatusCall(common.BackgroundContext(ctx))
+			now = time.Now()
 
-		var cacheObj runner.RunnerStatus
+			// Pointer store of 'cachePtr' is sufficient here as isWaiter/isCached above perform a shallow
+			// copy of 'cache'
+			pr.status.lock.Lock()
 
-		// A shallow copy is sufficient here, as we do not modify nested data in
-		// RunnerStatus in any way.
-		pr.status.lock.Lock()
+			pr.status.cache = cachePtr
+			pr.status.expiry = now.Add(StatusCallCacheDuration)
+			pr.status.wait = nil
 
-		cacheObj = *pr.status.cache
+			pr.status.lock.Unlock()
 
-		pr.status.lock.Unlock()
+			// signal waiters
+			close(waitChan)
+		}()
+	}
 
-		cacheObj.Cached = true
-		cacheObj.Active = atomic.LoadInt32(&pr.status.inflight)
-		cacheObj.RequestsReceived = atomic.LoadUint64(&pr.status.requestsReceived)
-		cacheObj.RequestsHandled = atomic.LoadUint64(&pr.status.requestsHandled)
-		return &cacheObj, nil
+	if !isCached {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-waitChan:
+		}
 	}
 
-	cachePtr := pr.runStatusCall(ctx)
-	cachePtr.Active = atomic.LoadInt32(&pr.status.inflight)
-	cachePtr.RequestsReceived = atomic.LoadUint64(&pr.status.requestsReceived)
-	cachePtr.RequestsHandled = atomic.LoadUint64(&pr.status.requestsHandled)
-	cachePtr.KdumpsOnDisk = atomic.LoadUint64(&pr.status.kdumpsOnDisk)
-	now = time.Now()
+	var cacheObj runner.RunnerStatus
 
-	// Pointer store of 'cachePtr' is sufficient here as isWaiter/isCached above perform a shallow
-	// copy of 'cache'
+	// A shallow copy is sufficient here, as we do not modify nested data in
+	// RunnerStatus in any way.
 	pr.status.lock.Lock()
 
-	pr.status.cache = cachePtr
-	pr.status.expiry = now.Add(StatusCallCacheDuration)
-	pr.status.wait = nil
+	cacheObj = *pr.status.cache
 
 	pr.status.lock.Unlock()
 
-	// signal waiters
-	close(myChan)
-	return cachePtr, nil
+	cacheObj.Cached = isCached
+	cacheObj.Active = atomic.LoadInt32(&pr.status.inflight)
+	cacheObj.RequestsReceived = atomic.LoadUint64(&pr.status.requestsReceived)
+	cacheObj.RequestsHandled = atomic.LoadUint64(&pr.status.requestsHandled)
+	cacheObj.KdumpsOnDisk = atomic.LoadUint64(&pr.status.kdumpsOnDisk)
```

Comment:
it maybe makes sense to do this only from the spawner goroutine and maybe before spawning the actual status call (as the context here itself could be indicative of why the status call fails), and return only that cached object instead of loading these fields at the current state for any cached object, then the status object returned is identical to the status when the status call itself ran for any calls to this where it's cached in a window. for example, the status call itself could run when there are 10000 outstanding requests, this could cause the status call to fail. while we're returning that failed state when returning a cached object, we're also going to reload the number of outstanding requests, which could now be 10. it makes sense when the status call fails with 10000 outstanding requests when maybe it makes a lot less sense when there are only 10 and also the server is in a different state wrt these than when the status call actually ran - at least, we've lost all the context of the outstanding status of the server when the status call failed (ie these fields) outside of what we get back from the status call itself. that was a bit rambly, if that's not coherent press 1 to try again...

Reply:
These fields are used by PM and they need to be up-to-date. Logically your comment makes sense as it's a bit weird here we keep some section of data up-to-date, but rest comes from cache. I'll add some comments to clarify why we are doing this odd data copy.